Built-in support for file and string comparisons

jgmdev opened the pull request:
This is a ~~work in progress~~ files and strings differences viewer, consist of a C backend for calculating the differences in a more performant way basically using the Longest Common Subsequence algorithm. The Lua side is basically the viewer. ~~The viewer is the last piece to finish, for now its just two DocViews side by side stuck inside a parent view that handles them.~~

Testing and improvements are welcome!

Implements feature request #305

### New Style Colors for Schemes

```lua
style.diff_delete = { common.color "rgba(200, 84, 84, 1)" }
style.diff_insert = { common.color "rgba(77, 127, 73, 1)" }
style.diff_modify = { common.color "rgba(202, 173, 85, 1)" }
style.diff_delete_background = { common.color "rgba(200, 84, 84, 0.30)" }
style.diff_insert_background = { common.color "rgba(77, 127, 73, 0.30)" }
style.diff_modify_background = { common.color "rgba(202, 173, 85, 0.30)" }
style.diff_delete_inline = { common.color "rgba(200, 84, 84, 0.40)" }
style.diff_insert_inline = { common.color "rgba(77, 127, 73, 0.40)" }
```

### New Commands

* diff-view:start-files-comparison
* diff-view:start-strings-comparison
* diff-view:prev-change
* diff-view:next-change
* diff-view:sync-change
* diff-view:select-file-for-compare
* diff-view:compare-file-with-selected
* diff-view:select-text-for-compare
* diff-view:compare-text-with-selected

### Preview

https://github.com/user-attachments/assets/f5856c26-be2b-44f4-8d68-73b472fb90e9

<img width="1920" height="1080" alt="diff viewer" src="https://github.com/user-attachments/assets/a8e5e42b-0688-41d5-8f4f-2f68acfdf482" />

<img width="774" height="507" alt="diff viewer settings" src="https://github.com/user-attachments/assets/8fbdf60c-43f7-4320-98da-44d58b1a231d" />

jgmdev committed:
Built-in support for file and string comparisons

Implements feature request #305

jgmdev committed:
Use a faster token based version of similarity

jgmdev committed:
try to fix msv errors

jgmdev committed:
Better identify indentical lines

jgmdev committed:
Some adjustments to the DiffView

jgmdev commented:
should be better now and a bit faster:

<img width="1148" height="813" alt="20250715-172447" src="https://github.com/user-attachments/assets/d8b16108-08f6-4c55-8cfd-e340498985ff" />

The View still has glitches and need of love but now supports scrolling both sides, we are getting somewhere :smiley: ... Right now one can edit directly on the views and save using bindings but no feature to re-calculate the differences yet. Also, due to the gap calculations between lines been a mess and missing DocView overrides, the position of the cursor is wrongly calculated.

jgmdev committed:
Add API doc file for diff

SwissalpS commented:
> do SDL 3 related commits do feature latest commits of File Comparison commits or not?

PR #254 (SDL 3 port) is fresher than this PR #306 and according to the commit tree they are separate branches.

jgmdev committed:
Fixes to the diff viewer.

* Now clicking takes you to the right line.
* Cleaned up the way docview overrides are done for less repetitive code.
* Fixed get visible lines range.
* Re-calculate differences on text insert or remove.
* Renamed [ab]_spaces to [ab]_gaps for more clarity

jgmdev committed:
Draw differences on scrollbar and sync arrows in  middle

jgmdev committed:
More improvements to the diff viewer

* Now supports syncing between A and B
* Added commands to navigate to next change and to apply them
  on the contrary view
* Sync indicators are now clickable
* Added a plugin API for usage by other plugins like scm.
* Fixed horizontal scrolling

jgmdev committed:
sync a and b scroll on diff update

jgmdev committed:
Fix out of range error

jgmdev committed:
Properly handle multiple modifies

* Fixed out of range error
* There is something that needs investigating when drawing inline diff

jgmdev committed:
Main scrollbar always expanded and 3 times bigger

* Fixed more out if sync scrollbar issues

jgmdev committed:
Added diff:start-file-comparison command

* When computing differences yield only every 0.5 seconds since the
  native differ is pretty fast now.
* Log the amount of time that it took to compute the differences.

jgmdev committed:
Fix out of place inline diff rendering

jgmdev commented:
> There's another small bug (not related to Minimap as I tested): if `Settings → Graphics → Animate Mouse Drag Scroll` is enabled, then dragging scrollbar or Minimap in Files Comparison mode results in uneven scrolling of the other file that is being compared.

Fixed, also made the main scrollbar expanded by default and bigger for easier mouse grabbing.

The 85kb files you shared before should be fully computed in 0.4 seconds (at least on my system). I was yielding for every line before and that is why it could take several seconds on long files. This means that now typing some characters reflects the changes almost immediately.

Should be in pretty good shape now (I would say good enough for merging) except for the scrolling bug which I haven't yet properly investigated what is causing it... Also, the colors, can't decide which style names to use (style.diff_{insert|delete|modify}, style.diff = {insert, delete, modify}), which approach to take since I also need to color the inline differences of modified lines... not sure which colors to use or if I should just disable syntax highlighting for lines with inserts and deletes for more readability...

jgmdev committed:
Added diff colors to default style

Also made showing the time it took to compute the differences
toggleable from the config options.

jgmdev committed:
Move main scrollbar to match doc views when scrolling a doc

jgmdev committed:
Added config for plain text rendering of changes

jgmdev commented:
> I spotted a small issue with the right most scrollbar (the third one): if you scroll by dragging file 1's or file 2's scrollbar, the third scrollbar won't move. If you use the third scrollbar, all three are moving correctly.

That is fixed, nice catch!

> Being able to customize the colors via theme would be perfect, as it's quite tricky to specify the color select the color that will look good in both dark and light color scheme.

That is possible now:

```lua
style.diff_delete = { common.color "rgba(200, 84, 84, 1)" }
style.diff_insert = { common.color "rgba(77, 127, 73, 1)" }
style.diff_modify = { common.color "rgba(202, 173, 85, 1)" }
style.diff_delete_background = { common.color "rgba(200, 84, 84, 0.30)" }
style.diff_insert_background = { common.color "rgba(77, 127, 73, 0.30)" }
style.diff_modify_background = { common.color "rgba(202, 173, 85, 0.30)" }
style.diff_delete_inline = { common.color "rgba(200, 84, 84, 0.40)" }
style.diff_insert_inline = { common.color "rgba(77, 127, 73, 0.40)" }
```

> Speaking of syntax highlighting enabled VS disabled: a toggle in plugin settings would be a perfect solution (personally, I'd like to use highlighting as when you use it your file looks as familiar as when you're working with it in regular mode).

Good idea! implemented and here are the previews:

<img width="1920" height="1080" alt="20250723-172122" src="https://github.com/user-attachments/assets/2f42f0fe-99b4-40fb-9c06-9f605b845c30" />

<img width="1920" height="1080" alt="20250723-172142" src="https://github.com/user-attachments/assets/9e919402-0dfc-42b4-987d-caf9f164af5f" />

Improves readability by a lot on any theme but... no syntax coloring when plain text is enabled :)

<img width="774" height="507" alt="20250723-172803" src="https://github.com/user-attachments/assets/8fbdf60c-43f7-4320-98da-44d58b1a231d" />

jgmdev committed:
Fix docview top selection

jgmdev committed:
Added diff:start-strings-comparison command

Also string comparisons are now unnamed to allow user to choose
when saving.

jgmdev committed:
Renamed the diff plugin to diffview

This prevents clashing with the core diff library.

Also, added the option of been able to name a or be when
dealing with a string comparison.

jgmdev committed:
Added overload doc comment to DiffView

jgmdev committed:
More doc comments goodness

jgmdev committed:
Fix line 1 syncing

jgmdev committed:
Made command plural

jgmdev commented:
Fixed the selection bug, renamed the plugin from diff to diffview to prevent clashes with native diff library. Will proceed to merge this, any remaining bugs may be found on the daily usage.

What is left is a good samaritan to update the default and color repo syntax files to have sane diff colors.
